src/security/hapi-auth-validation.js

```diff
@@ -1,68 +1,134 @@
+  const resourceBuilder = authConfig.resources[resourceType]
+
+  if (!resourceBuilder) {
+    return done({ error: 'Resource builder not found' })
+  }
+
+  const requestParams = plugin.getParams ? plugin.getParams(request) : {}
+  const buildParams = Object.assign({}, { organizationId }, requestParams)
+
+  if (resourceType === 'users' && buildParams.userId) {
+    return buildResourcesForUser(resourceBuilder, buildParams.userId, organizationId, done)
+  }
+
+  done(null, [resourceBuilder(buildParams)])
+}
+
+function authorize (job, next) {
+  if (!job.plugin) return next(Boom.forbidden('Invalid credentials', 'udaru'))
+
+  buildResources(job.plugin, job.request, job.organizationId, (err, resources) => {
+    if (err) return Boom.unauthorized('Bad credentials')
+
+    const action = job.plugin.action
+    const userId = job.currentUser.id
+
+    async.any(resources, async.apply(checkAuthorization, userId, action), (err, valid) => {
+      if (err) return next(Boom.forbidden('Invalid credentials', 'udaru'))
+      if (!valid) return next(Boom.forbidden('Invalid credentials', 'udaru'))
+
+      next()
+    })
+  })
+}
 
 module.exports = (options, server, request, userId, callback) => {
-  async.waterfall([
-    (next) => {
-      userOps.getUserOrganizationId(userId, (error, organizationId) => {
-        if (error) {
-          return next({ message: 'Bad credentials' })
-        }
-
-        next(null, organizationId)
-      })
-    },
-    (organizationId, next) => {
-      userOps.readUser({ id: userId, organizationId }, (error, user) => {
-        if (error) {
-          return next({ message: 'Bad credentials' })
-        }
-
-        next(null, user)
-      })
-    },
-    (user, next) => {
-      const { route } = request
-
-      const authPlugin = route.settings.plugins && route.settings.plugins.auth
-      if (!authPlugin) {
-        return next(null, false)
-      }
-
-      const requestParams = authPlugin.getParams ? authPlugin.getParams(request) : {}
-      const buildParams = Object.assign({}, user, requestParams)
-
-      let resource = authPlugin.resource
-      if (!resource) {
-        const resourceType = route.path.split('/')[2]
-        if (!authConfig.resources[resourceType]) {
-          return callback(null, false, { error: 'Resource builder not found' })
-        }
-
-        const resourceBuilder = authConfig.resources[resourceType]
-        resource = resourceBuilder && resourceBuilder(buildParams) || null
-      }
-
-      const params = {
-        userId,
-        action: authPlugin.action,
-        resource
-      }
-
-      authorize.isUserAuthorized(params, (error, result) => {
-        if (error) {
-          return next(error)
-        }
-
-        next(null, result.access, user)
-      })
-
-    }
-  ], callback)
+
+  const job = {
+    userId,
+    request,
+    requestedOrganizationId: request.headers.org,
+    plugin: request.route.settings.plugins && request.route.settings.plugins.auth
```

Comment:
could `plugin` be rename to `authorization`/`authorizationParams` or something similar? I think `plugin` is a bit too generic as a name in this context